core/src/bls_sigverify/sigverifier.rs

```diff
@@ -0,0 +1,1681 @@
+        }
+    }
+
+    // Given a vote message, following preprocessing steps are performed:
+    // 1. Get the corresponding vote account pubkey and BLS pubkey from rank.
+    // 2. Capture metrics for received votes.
+    // 3. Check if the vote is newer than root slot, discard if not.
+    // Uppon success, returns the vote account pubkey and BLS pubkey.
+    fn preprocess_vote(
+        &mut self,
+        vote_message: VoteMessage,
+        root_bank: &Bank,
+        consensus_metrics_to_send: &mut Vec<ConsensusMetricsEvent>,
+    ) -> Option<(Pubkey, BlsPubkey)> {
+        // Missing epoch stakes
+        let Some(key_to_rank_map) = get_key_to_rank_map(root_bank, vote_message.vote.slot()) else {
+            self.stats
+                .received_no_epoch_stakes
+                .fetch_add(1, Ordering::Relaxed);
+            return None;
+        };
+
+        // Invalid rank
+        let Some((vote_account_pubkey, bls_pubkey)) =
+            key_to_rank_map.get_pubkey(vote_message.rank.into())
+        else {
+            self.stats.received_bad_rank.fetch_add(1, Ordering::Relaxed);
+            return None;
+        };
+
+        // Capture votes received metrics before old messages are potentially discarded below.
+        consensus_metrics_to_send.push(ConsensusMetricsEvent::Vote {
+            id: *vote_account_pubkey,
+            vote: vote_message.vote,
+        });
+        // Only need votes newer than root slot
+        if vote_message.vote.slot() <= root_bank.slot() {
+            self.stats.received_old.fetch_add(1, Ordering::Relaxed);
+            return None;
+        }
+        Some((*vote_account_pubkey, *bls_pubkey))
+    }
+
+    fn preprocess_packets(
+        &mut self,
+        packet_batches: &mut [PacketBatch],
+        root_bank: &Bank,
+        consensus_metrics_to_send: &mut Vec<ConsensusMetricsEvent>,
+    ) -> (Vec<VoteToVerify>, Vec<Certificate>) {
+        // TODO(sam): ideally we want to avoid heap allocation, but let's use
+        //            `Vec` for now for clarity and then optimize for the final version
+        let mut votes_to_verify = Vec::with_capacity(AVERAGE_VOTES_OR_CERTS_PER_BATCH);
+        let mut certs_to_verify = Vec::with_capacity(AVERAGE_VOTES_OR_CERTS_PER_BATCH);
+        for mut packet in packet_batches.iter_mut().flatten() {
+            self.stats.received.fetch_add(1, Ordering::Relaxed);
+            if packet.meta().discard() {
+                self.stats
+                    .received_discarded
+                    .fetch_add(1, Ordering::Relaxed);
+                continue;
+            }
+
+            let message: ConsensusMessage = match packet.deserialize_slice(..) {
+                Ok(msg) => msg,
+                Err(_) => {
+                    self.stats
+                        .received_malformed
+                        .fetch_add(1, Ordering::Relaxed);
+                    packet.meta_mut().set_discard(true);
```

Comment:
For my education: The caller of this function takes ownership of batches and after calling this function drops them.  So why do we need to set discard on the packets here?

Reply:
This was carved out of the ed25519 Sigverifier, so we kept most of the original structure intact. I believe the original intent was to do some statistics, so it needs this discard flag.

However, we are thinking about dropping the sigverify_service and remove the deduper, so this will be gone soon.